### PR TITLE
Introduce a caching layer for Image get requests

### PR DESCRIPTION
**Pull request by Piojo**
During some executions we may instantiate multiple `Image` objects that point to the same image. Operations like fetching their manifests incurs in usage of API quota on some registries, so it's cheaper for everyone to issue a `HEAD` request first and see if we have it in a cache. If we do, we return that cache, if not, we issue the `GET` request and cache its contents.

For backwards compatibility, this caching is optional, and we would incur in storing a single-key dictionary in such case.

The key for the cache will be both the `etag`, if there is one, or the image's digest. Not all registries return etags for their objects, but  those that do use the digest as the etag, so in that particular case they are interchangeable.

### CODE REVIEW

**Comment by maorfr** [comment]
related to https://issues.redhat.com/browse/APPSRE-3385

File: sretoolbox/container/image.py
Comment (apahim) [inline sretoolbox/container/image.py:223]: There's a circular dependency here. `manifest()` calls `digest()` that calls `manifest()`. The cache should be ETag based, not digest based.

Comment (Piojo) [reply sretoolbox/container/image.py:226]: Dang, I had missed the `or self.digest` part in `_fetch_manifest`. I'll break that dependency in a moment.

Comment (Piojo) [reply sretoolbox/container/image.py:226]: ... actually no. If you check the `test_parser` test, it calls `digest` on each of the images and we don't enter any kind of recursive loop or call cycle.

Comment (apahim) [reply sretoolbox/container/image.py:226]: great, tks for checking

File: sretoolbox/container/image.py
Comment (apahim) [inline sretoolbox/container/image.py:102]: if we are not doing the etag thing, this should probably be renamed

File: sretoolbox/container/image.py
Comment (apahim) [inline sretoolbox/container/image.py:199]: Both digests and etags as keys to the same cache? Is this intentional or a leftover?
Isn't this cached already at the request level?

Comment (Piojo) [reply sretoolbox/container/image.py:199]: It is intentional - it may be already cached at the request level, if the registry did return an etag. Or it may need caching here if the registry doesn't.
If there is an etag, it most likely will be identical to the digest - in the rare registry that doesn't, oh, well, we've wasted 10KB a few times.

**Review by apahim — CHANGES_REQUESTED** [review]
We had that dicussion already in the design doc, but I also want to get this moving, so I propose you split this into 2 PRs, the second one containing only the digest-based cache implementation. That makes sense anyway, since the digest-based cache is not related to the Story you're working on.

File: sretoolbox/container/image.py
Comment (apahim) [reply sretoolbox/container/image.py:199]: got it. tks for clarifying.

**Comment by Piojo** [comment]
The digest/manifest-specific cache is 6 lines of code, and splitting the PRs would make them depend on each other, which is confusing and a mess to manage.

I agree that the history had gotten messy, and needed a cleanup before merging. It is now two commits, one for the overall cache and another one for the manifest-specific code.

**Comment by apahim** [comment]
Hey @Piojo, welcome back :)

Having a PR depending on another is not confusing or messy. It's actually simple and straight forward, both git and github can manage that well.

But the reason I'm asking you to split is not related to commit organization, even though I do appreciate it. The reason I propose the split is that those 6 lines introduce a bad system design, they are not related to the Story you are working on and they are a point of disagreement since the design doc, they should not even have showed up here. So far, you did not manage to convince anyone we need them.

So I'm giving you a path forward to complete the Story you are implementing. I was under the assumption it was an important one. Right now, this PR is blocked on that, but the split would unblock the part that is relevant for the Story you are implementing, while allowing more time for you to convince us we need those unrelated 6 lines. Of course, your call.

File: sretoolbox/container/image.py
Comment (apahim) [inline sretoolbox/container/image.py:388]: this will likelly change with the previous suggestion, but I'm not convinced we need to import `contextlib` just to have a shorter try/except block.

File: sretoolbox/container/image.py
Comment (apahim) [inline sretoolbox/container/image.py:393]: we did a lot before even knowing if we should cache. This should probably be the first thing to check before hitting the cache backend or doing a HEAD request.

File: tests/test_container.py
Comment (apahim) [inline tests/test_container.py:16]: Nitpick... this sintax is good according to PEP8, but it's not diff friendly. I'd suggest not to group the imports. No big deal though.

File: tests/test_container.py
Comment (apahim) [inline tests/test_container.py:301]: Thank you for the tests... they are likelly to change, so I will wait the next update to review them.

File: sretoolbox/container/image.py
Comment (apahim) [inline sretoolbox/container/image.py:385]: Before making this request, It's cheaper to first check if we have a cache record for that url.

**Review by apahim — CHANGES_REQUESTED** [review]
Hi @Piojo, thank you for splitting the PR! I love the refactorings and the tests. I left some minor comments, but I have some concerns about the overall performance... I believe we can avoid many calls by first checking if we should be caching at all, then checking if we have a cached data for the url wa want, and only then do the HEAD call. Otherwise, we can do the GET call directly. Please let me know if that makes sense to you.

File: sretoolbox/container/image.py
Comment (apahim) [inline sretoolbox/container/image.py:373]: same here... before any `head` request, we should check the cache

File: sretoolbox/container/image.py
Comment (apahim) [inline sretoolbox/container/image.py:391]: you're passing `auth` toghether with `Authorization` headers. Not sure how safe this is.

File: sretoolbox/container/image.py
Comment (apahim) [inline sretoolbox/container/image.py:82]: If a cache backend is bot provided, we should not cache at all. The cache is only useful when shared between instances

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:385]: Not really, because the object an URL points to can change. For instance, when operating with a `latest` image, that tag can move. My cache key is the etag - that's what we've been discussing all around - not the URL.

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:82]: Yes. But as I mention in the description the code is way simpler if I fake it locally and keep a one-entry cache, instead of polluting the consumers with awful `if self.response_cache` blocks.

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:360]: ... simple is better than complex said the Zen of Python, so unless you can give me a simpler expression, I'll keep this form.

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:72]: It is a class variable, it will only be computed when the class is first loaded. 50KB is an arbitrary number - most manifests I'm seeing are on the order of 10-15KB, so 50KB gives us enough room for larger instances.

I don't want to store arbitrarily large items in the cache, to protect us from OOM-ing too easily.

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:388]: See my response to that suggestion.

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:371]: Sure.

File: sretoolbox/container/image.py
Comment (apahim) [reply sretoolbox/container/image.py:73]: Awesome, please capture that info a comment

File: sretoolbox/container/image.py
Comment (apahim) [reply sretoolbox/container/image.py:361]: It's just a suggestion. Tks for considering.

File: sretoolbox/container/image.py
Comment (apahim) [reply sretoolbox/container/image.py:386]: If there's nothing in the cache, no need for HEAD requests.

File: sretoolbox/container/image.py
Comment (apahim) [reply sretoolbox/container/image.py:83]: `if` is cheaper than `request.head`. Way chepaer.

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:392]: So far I see it working, if you have a scenario that worries you I'll make sure to test it and get it fixed.

Comment (apahim) [reply sretoolbox/container/image.py:392]: Yes, you are setting the `Authorization` header twice, with different data. Why are you doing that? What's the purpose of having it set twice, with different data?

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:386]: I disagree - as of today, the API for the cache states the key is the `etag` - this is what was agreed on the design document. Therefore there is no better-than-O(N) way to look up a given URL. This is also what prompted the separation of the manifest-specific caching, let's not move the goal posts.

Then, this is also assuming that it isn't possible for objects to get renamed, that there will never be any redirects or that different mirrors won't hold the same objects. I doubt those assumptions belong in a simple wrapper around HTTP `GET`.

If we decided etags rule (and for good reason), then we have to fetch the etag first.

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:394]: To keep the code simple and avoid branching based on whether a header has been inserted somewhere. I've updated the 401 block to discard the `auth` tuple if it isn't valid for this image.

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:83]: 1. Performance doesn't seem to be a problem in here. GET request quota is.
2. Show me a way around `head` without conditional requests, which aren't implemented in our most pressing use case.

File: sretoolbox/container/image.py
Comment (apahim) [reply sretoolbox/container/image.py:386]: We are mixing a bit the topics here. But it's fine, we can always improve, if/when performance is an issue. No need to improve for performance beforehand, if that's not in line with the design you believe in.
Thanks for taking it into consideration anyway.

File: sretoolbox/container/image.py
Comment (apahim) [reply sretoolbox/container/image.py:390]: How is the `contextlib.suppress` better than a try/except block? Is that just for readability?

File: sretoolbox/container/image.py
Comment (apahim) [reply sretoolbox/container/image.py:83]: 1. Performance doesn't seem to be a problem in here. GET request quota is.

You're making a lot of useless requests. To solve your use-case, you're penalizing whoever else is using this lib, when you don't really have to. This is a community project, we should assume there are more users than qontract-reconcile. Even internally, many other projects are using this lib.

2. Show me a way around head without conditional requests, which aren't implemented in our most pressing use case.

```
if self.response_cache is None:
```

File: sretoolbox/container/image.py
Comment (apahim) [inline sretoolbox/container/image.py:367]: Why re we reducing from 5 to 1 max_attempts?

Comment (Piojo) [reply sretoolbox/container/image.py:367]: Mistake from another test, sorry about this one.

File: sretoolbox/container/image.py
Comment (Piojo) [reply sretoolbox/container/image.py:390]: Yes, it's a readability improvement.

**Review by apahim — APPROVED** [review]
Looking quite good, thank you for the updates. A few comments pending resolution, we are good to go after them.

**Pull request merged**
